src/main/java/com/google/devtools/build/lib/rules/cpp/CppLinkAction.java

```diff
@@ -466,20 +485,118 @@ protected String getRawProgressMessage() {
+      this.inputs = inputs;
+    }
+
+    /** Performs costly computations required to predict linker resources consumption. */
+    private LazyData doCostlyEstimation() {
+      int inputsCount = 0;
+      long inputsBytes = 0;
+      for (Artifact input : inputs.toList()) {
+        inputsCount += 1;
+        try {
+          FileArtifactValue value =
+              actionExecutionContext.getInputMetadataProvider().getInputMetadata(input);
+          if (value != null) {
+            inputsBytes += value.getSize();
+          } else {
+            throw new IOException("no metadata");
+          }
+        } catch (IOException e) {
+          // TODO(https://github.com/bazelbuild/bazel/issues/17368): Propagate as ExecException when
+          // input sizes are used in the model.
+          logger.atWarning().log(
+              "Linker metrics: failed to get size of %s: %s (ignored)", input.getExecPath(), e);
+        }
+      }
+
+      // TODO(https://github.com/bazelbuild/bazel/issues/17368): Use inputBytes in the computations.
+      ResourceSet resourceSet;
+      switch (os) {
+        case DARWIN:
+          resourceSet =
+              ResourceSet.createWithRamCpu(
+                  /* memoryMb= */ 15 + 0.05 * inputsCount, /* cpuUsage= */ 1);
+          break;
+        case LINUX:
+          resourceSet =
+              ResourceSet.createWithRamCpu(
+                  /* memoryMb= */ Math.max(50, -100 + 0.1 * inputsCount), /* cpuUsage= */ 1);
+          break;
+        default:
+          resourceSet =
+              ResourceSet.createWithRamCpu(/* memoryMb= */ 1500 + inputsCount, /* cpuUsage= */ 1);
+          break;
+      }
+
+      return new LazyData(inputsCount, inputsBytes, resourceSet);
+    }
+
+    @Override
+    public ResourceSet get() throws ExecException {
+      if (lazyData == null) {
+        lazyData = doCostlyEstimation();
+      }
+      return lazyData.resourceSet;
+    }
+
+    /**
+     * Emits a log entry with the linker resource prediction statistics.
+     *
+     * <p>This should only be called for spawns where we have actual linker consumption metrics so
+     * that we can compare those to the prediction. In practice, this means that this can only be
+     * called for locally-executed linkers.
+     *
+     * @param actualMemoryKb memory consumed by the linker in KB
+     */
+    public void logMetrics(long actualMemoryKb) {
+      if (lazyData == null) {
+        // We should never have to do this here because, today, the memory consumption numbers in
+        // actualMemoryKb are only available for locally-executed linkers, which means that get()
+        // has been called beforehand. But this does not necessarily have to be the case: we can
+        // imagine a remote spawn runner that does return consumed resources, at which point we
+        // would get here but get() would not have been called.
+        lazyData = doCostlyEstimation();
+      }
+
+      logger.atInfo().log(
```

Comment:
Late to the party - do you need this log for *every* link action or is it ok to get a sample? I am asking because this seems to cause a bit of log spam when looking at large builds.

Reply:
Oops, sorry for the late reply.

Yes, I kinda needed it for every link action. If I look at our build, there are lots and lots of small link actions, but towards the end of the build, there are only a few large ones, each with a different memory "profile". If I were to sample these, I'd lose valuable information at that stage.

Is this still a problem after @lberki 's change to decrease the logging level?